feat:sparkles: Create chatrooms for players

gchamb opened the pull request.

gchamb committed:
feat(client): 🚧 Logic for Chatting

- some logic for chat and chatrooms
- create chatroom function
- added chatroom collection

gchamb committed:
feat(client): ✨ Functional logic for chatroom

- placed appropriate code in server
- create get collection functions
- subscribe the chat messages

gchamb committed:
Merge branch 'development' into giantte-muaad/chat

gchamb committed:
Merge branch 'development' into giantte-muaad/chat

gchamb committed:
Merge branch 'development' into giantte-muaad/chat

gchamb committed:
feat(client): ✨ Implement Basic Chat Features

- Able to enter chatrooms
- Able to send messages back and forth
- Fix collection for chatmessages
- dummy data for slug route(temporary)

gchamb committed:
feat(client): ✨ Functional Chat Room

- timer on top
- display who you matched with
- dummy styles for now
- chat functions that will be moved over to the server in the future

gchamb committed:
Merge branch 'development' into giantte-muaad/chat

gchamb committed:
feat(functions): ✨ Switch client functions into server functions

- switched logic to server functions
- add different collections

gchamb committed:
feat(functions): ✨ Secure adding messages

- remove messaging from client
- created another request validator

gchamb committed:
refactor(functions): ♻️ comment out stalker for now

gchamb committed:
refactor(client):

gchamb committed:
fix(functions): 🐛 linting errors for functions

gchamb committed:
feat(functions): ✨ Return stalker for use

gchamb committed:
Merge branch 'development' into giantte-muaad/chat

gchamb committed:
feat(firestore): create firestore rule

- create firestore to allow people to read from their chatroom and messages

gchamb committed:
refactor(client): 🔥 remove chat component

nstringham commented:
can you set the pull request name to something that describes the entire pull request?

gchamb commented:
> can you set the pull request name to something that describes the entire pull request?

You like the new name? 😏

nstringham commented:
> > can you set the pull request name to something that describes the entire pull request?
> 
> You like the new name? 😏

This PR includes files in client and server so it doesn't seam like the `firestore` scope applies here

nstringham commented on client/src/lib/firebase/firestore-collections.ts:
```diff
@@ -1,16 +1,21 @@
 import { collection, type CollectionReference, type DocumentReference } from "firebase/firestore";
 import { db } from "$lib/firebase/app";
 
+export const lobbyCollection = collection(db, "lobbies") as CollectionReference<Lobby>;
```
I think it made sense for the userCollection to be above the lobbyCollection so thet all the lobby stuff was together.

nstringham commented on client/src/lib/firebase/firestore-collections.ts:
```diff
@@ -1,16 +1,21 @@
 export function getPrivatePlayerCollection(lobbyDoc: DocumentReference<Lobby>): CollectionReference<PrivatePlayer> {
   return collection(lobbyDoc, "privatePlayers") as CollectionReference<PrivatePlayer>;
 }
-
```
I liked the blank line but I guess it can go away :(

nstringham commented on functions/src/firestore-collections.ts:
```diff
@@ -11,6 +12,12 @@ export function getPrivatePlayerCollection(lobbyDoc: DocumentReference<Lobby>):
   return lobbyDoc.collection("privatePlayers") as CollectionReference<PrivatePlayer>;
 }
 
+export function getChatRoomCollection(lobbyDoc: firestore.DocumentReference<Lobby>) {
+  return lobbyDoc.collection("chatRooms") as firestore.CollectionReference<ChatRoom>;
+}
+export function getChatRoomMessagesCollection(roomDoc: firestore.DocumentReference<ChatRoom>) {
+  return roomDoc.collection("chatMessages") as firestore.CollectionReference<ChatMessage>;
+}
```
I think  this file should be in the same order as the one in client

nstringham commented on functions/src/firestore-collections.ts:
```diff
@@ -1,7 +1,8 @@
-import type { Lobby, PrivatePlayer, PromptAnswer } from "./firestore-types/lobby";
+import { ChatMessage, ChatRoom, Lobby, PrivatePlayer, PromptAnswer } from "./firestore-types/lobby";
+import { db } from "./app";
+import { firestore } from "firebase-admin";
```
let's use the import from "firebase-admin/firestore" instead of the firestore namespace for types

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,62 @@
+});
+
+export const deleteChatRooms = functions.https.onCall(async (data: unknown, context) => {
+  if (!context.auth) {
```
Are we sure we want any logged in user to be able to delete the chat rooms in any lobby?

nstringham commented on firestore.rules:
```diff
@@ -19,6 +19,13 @@ service cloud.firestore {
+      match /chatRooms/{roomId} {
+        allow read: if true;
+        match /chatMessages/{messageId}{
+          allow read: if true;
```
we definitly don't want anyone to be able to see any chat

nstringham commented on firestore.rules:
```diff
@@ -19,6 +19,13 @@ service cloud.firestore {
       }
+
+      match /chatRooms/{roomId} {
+        allow read: if true;
```
I don't think we want to allow everyone to view who is paired with who

nstringham replied:
still an issue

nstringham commented on client/src/routes/game/+page.svelte:
```diff
@@ -10,6 +9,7 @@
   import { authStore as user } from "$stores/auth";
   import { goto } from "$app/navigation";
   import type { Unsubscribe } from "firebase/auth";
+    import ChatRoom from "$components/ChatRoom.svelte";
```
lets put this with the other components

nstringham commented on client/src/routes/game/+page.svelte:
```diff
@@ -60,7 +60,7 @@
     unsubscribePrivatePlayer = onSnapshot(privatePlayerDocRef, (doc) => {
       // will change privatePlayer to the new doc data
-      privatePlayer = doc.data();
+      privatePlayer = doc.data() as PrivatePlayer;
```
let's not have an unnessisary `as`

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,161 @@
+    chatRoomInfo = await findChatRoom(lobbyData.id, user.uid);
+    // subscribe the chat messages
+    onSnapshot(getChatRoomMessagesCollection(lobbyData.id, chatRoomInfo.id), (collection) => {
+      chatMessages = collection.docs.map((doc) => doc.data()).sort((a, b) => a.timestamp.seconds - b.timestamp.seconds);
```
why don't you put an "orderby" in the querry so that you don't have to sort on the client?

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,161 @@
+  export let lobbyData: Lobby & { id: string };
+  // variables
+  let user = $authStore as User;
+  let userInfo: UserData;
+  let partnerInfo: UserData;
```
these should be type `Player` not `UserData`

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,161 @@
+    });
+    partnerInfo = lobbyData.players[lobbyData.uids.indexOf(partner)];
+    // create timer
+    timer = setInterval(() => {
+      countdown -= 1;
+    }, 1000);
```
this is not a reliable way to do a countdown.
if you're interested here is an exploration into javascript timer techniqes https://www.youtube.com/watch?v=MCi6AZMkxcU

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,161 @@
+  // Function will create document with new message
+  function submitMessage() {
+    if (message === "") {
+      return;
```
should also probobly disable the button

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,161 @@
+  $: if (countdown === 0) {
+    clearInterval(timer);
+    // Deletes all the chatrooms and switches game state
+    deleteChatRooms({ code: lobbyData.id });
```
what happens if more than one client calls this function?

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,161 @@
+    <div>MATCHED WITH {partnerInfo.displayName.toUpperCase()}</div>
+  {/if}
+  <div class="messages">
+    {#if chatMessages.length > 0}
```
this if is probobly not nessary because it is the same as looping zero times

nstringham commented on functions/src/index.ts:
```diff
@@ -1 +1,5 @@
 export * from "./lobby";
+export * from "./chat";
+
+// Start writing Firebase Functions
+// https://firebase.google.com/docs/functions/typescript
```
don't need this comment

nstringham commented on functions/src/lobby.ts:
```diff
@@ -165,7 +168,25 @@ export const collectPromptAnswers = functions.firestore
       transaction.update(lobbyDocRef, { state: "CHAT" });
 
       functions.logger.log(promptAnswers);
```
we can remove this line now that we are doing something with promptAnswers. it was just here for debugging

nstringham commented on functions/src/util.ts:
```diff
@@ -0,0 +1,46 @@
+      // generate a random player if the current pairs are equal
+      while (pairOne === pairTwo) {
+        pairTwo = lobbyData.uids[Math.floor(Math.random() * lobbyData.uids.length)];
+      }
```
this can be a do while loop to reduce code duplication

nstringham commented on functions/src/util.ts:
```diff
@@ -0,0 +1,46 @@
+
+export function generatePairs(lobbyData: Lobby): { pairs: { one: string; two: string }[]; stalker?: string } {
+  const pairs: { one: string; two: string }[] = [];
+  let pairOne, pairTwo;
```
these should have types

nstringham replied:
still need types

nstringham commented on functions/src/util.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+  // even number of players
+  else {
+    while (lobbyData.uids.length !== 0) {
+      // get random pair
+      pairOne = lobbyData.uids[Math.floor(Math.random() * lobbyData.uids.length)];
+      pairTwo = lobbyData.uids[Math.floor(Math.random() * lobbyData.uids.length)];
+      // generate a random player if the current pairs are equal
+      while (pairOne === pairTwo) {
+        pairTwo = lobbyData.uids[Math.floor(Math.random() * lobbyData.uids.length)];
+      }
+      // organize the pairs
+      pairs.push({ one: pairOne, two: pairTwo });
+      // delete them from temp array
+      lobbyData.uids.splice(lobbyData.uids.indexOf(pairOne), 1);
+      lobbyData.uids.splice(lobbyData.uids.indexOf(pairTwo), 1);
+    }
```
you can have to loop condition be `lobbyData.uids.length > 1` so that you don't need to have a seprate case for even or odd players. that way we can reduce code duplication.

gchamb committed:
style(client): 🎨 fix collection order

fixed collection order in firestore collections in client and server

gchamb committed:
fix(functions): 🔒️ dont allow anyone to see chatroom and messages

- firestore rules for chat and only let host delete the chatrooms

gchamb committed:
style(functions): 🎨 Add Type and Remove Unneccesary lines

gchamb committed:
refactor(functions): ♻️ remove code duplication

gchamb committed:
fix(client): 🐛 Better timer and minor changes

gchamb committed:
style(functions): 🎨 prettier fix

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,162 @@
+    timer = setInterval(() => {
+      const diff = Math.floor((end - Date.now()) / 1000);
+      countdown = diff;
+    }, 1000);
```
might want to consider setting this to a smaller number

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,162 @@
+    chatRoomInfo = await findChatRoom(lobbyData.id, user.uid);
+    // subscribe the chat messages
+    onSnapshot(getChatRoomMessagesCollection(lobbyData.id, chatRoomInfo.id), async (collection) => {
+      const messages = await getDocs(query(collection.query, orderBy("timestamp", "asc")));
```
this does a duplicater read. you want to put the querry itself into onSnapshot

nstringham commented on functions/src/util.ts:
```diff
@@ -0,0 +1,30 @@
+  while (lobbyData.uids.length > 1) {
+    // get random pair
+    pairOne = lobbyData.uids[Math.floor(Math.random() * lobbyData.uids.length)];
+    pairTwo = lobbyData.uids[Math.floor(Math.random() * lobbyData.uids.length)];
```
this line can be removed because of the do while loop

gchamb replied:
we still need the while loop because if not it will only generate one pair

nstringham reviewed (dismissed):
it seams like a few of my review comments from last time got marked as resolved even no resolution was reached

gchamb committed:
fix(firestore): 🔒️ Dont allow anyone to read chatroom

gchamb committed:
fix(client): 🐛 query in onsnapshot for messages and shorten interval

gchamb committed:
refactor(functions): ♻️ add type to pairs variable

gchamb committed:
Merge branch 'giantte-muaad/chat' of https://github.com/cat-acronym-team/cat-me-if-you-can into giantte-muaad/chat

gchamb committed:
fix(firestore): 🐛 only allow host to delete chatrooms

gchamb committed:
fix(functions): 🐛 have a co host who can delete as well

altered firebase rule to allow co host to delete and same for callable function

gchamb committed:
refactor(client): ♻️ minor changes for a way i like better

- switched object error to a string

gchamb committed:
Merge branch 'development' into giantte-muaad/chat

gchamb committed:
refactor(client): ♻️ move addmessage to client with firestore rule

- moved add message to client
- created firestore rule for it

gchamb committed:
feat(functions): ✨ add updated firebase admin

gchamb committed:
feat(functions): test enqueue and delete chatrooms

- turned delete chatrooms into a regular function
- testing enqueur in lobby update and chat
- added lobbychatmessages collection

gchamb committed:
fix(functions): 🐛 only alive players get paired

gchamb committed:
feat(client): ✨ Adjust timer to verify expiration

- add expiration attribute to lobby type

- created verify expiration to check if they submitted the correct time

- adjust the frontend to use expiration

gchamb committed:
refactor(functions): ♻️ remove cloud task function, chat request type, and add expiration request

gchamb committed:
style(functions): 🎨 minor prettier adjustment

gchamb committed:
Merge branch 'development' into giantte-muaad/chat

gchamb committed:
style(functions): 🎨 fix prettier and linting

gchamb committed:
fix(functions): 🐛 fix unresolved path

removed timestamp import and use the already imported firestore

gchamb committed:
refactor(client): ♻️ use find instead of filter

zaverbac reviewed (dismissed):
Everything looks good.

nstringham committed:
style(firestore): 🎨 add whitespace

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,163 @@
+    // subscribe the chat messages
+    onSnapshot(
+      query(getChatRoomMessagesCollection(lobbyData.id, chatRoomInfo.id), orderBy("timestamp", "asc")),
+      async (collection) => {
```
async without await

nstringham commented on firestore.rules:
```diff
@@ -19,6 +25,34 @@ service cloud.firestore {
+              && request.resource.data.timestamp is timestamp && request.resource.data.timestamp == request.time;
+          }
+          allow read: if canRead();
+          allow write: if inPair() && validateChatMessage();
```
should be `create`

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,25 @@
+import { Lobby } from "./firestore-types/lobby";
+
+export function deleteChatRooms(lobbyDoc: DocumentReference<Lobby>) {
+  db.runTransaction(async (transaction) => {
```
should return or await the runTransaction promise

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,25 @@
+
+export function deleteChatRooms(lobbyDoc: DocumentReference<Lobby>) {
+  db.runTransaction(async (transaction) => {
+    const lobbyDocRef = (await transaction.get(lobbyDoc)).ref;
```
?

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,25 @@
+    const chatRooms = await getChatRoomCollection(lobbyDocRef).listDocuments();
+
+    // get prompts answer then delete chatrooms
+    const lobbyChatCollection = getLobbyChatCollection(lobbyDocRef);
```
as we discussed lets place this into the lobby doc and put a isPromptAnswer (or smething like that) property on chat message

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,25 @@
+
+    // get prompts answer then delete chatrooms
+    const lobbyChatCollection = getLobbyChatCollection(lobbyDocRef);
+    chatRooms.forEach(async (room) => {
```
this async never gets awaited

nstringham commented on client/src/lib/firebase/chat.ts:
```diff
@@ -0,0 +1,36 @@
+  // Make sure sender is in the pair
+  if (!room.data().pair.includes(sender)) {
+    throw new Error("You can't send messages!");
+  }
```
lets not do this database read every time they send a message

nstringham commented on functions/src/firestore-functions-types.ts:
```diff
@@ -1,8 +1,33 @@
 export type ErrorResponse = { error?: string };
 
+export type ExpirationRequest = {
+  time: Timestamp;
```
lets not trust the client to tell us what time it is

nstringham commented on functions/src/firestore-functions-types.ts:
```diff
@@ -1,8 +1,33 @@
+export type GeneratedPairs = {
+  pairs: { one: string; two: string }[];
+  stalker?: string;
+};
```
lets keep this file for only things that are relivant to the client

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,163 @@
+    // Get userInfo
+    userInfo = lobbyData.players[lobbyData.uids.indexOf(user.uid)];
+    // Get partnerInfo
+    const partner = chatRoomInfo.data().pair.find((u) => {
```
no one letter variable names sir 😠

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,163 @@
+      // clear the input
+      message = "";
+      // if there's an error message then clear it
+      if (errorMessage !== "") {
+        errorMessage = "";
+      }
```
we don't really need this comment

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,163 @@
+    clearInterval(timer);
+    verifyExpiration({ time: Timestamp.now(), code: lobbyData.id });
+  }
+  $: if (countdown < 0) {
```
lets make that a -5

gchamb committed:
fix(firestore): 🐛 change allow write to allow create

gchamb committed:
refactor(client): ♻️ remove client time and type

- remove client request with time
- request expiration request
- remove unneeded validation
- fix mini mistakes

gchamb committed:
fix(functions): 🐛 return transaction and promise all

gchamb committed:
feat(functions): 🐛 added atrribute to lobby and player

- added new attributes to lobby and player
- adjusted verifyexpiration to use date.now()

gchamb committed:
refactor(functions): ♻️ relocate pairs type and minor refactor

nstringham commented on functions/src/util.ts:
```diff
@@ -0,0 +1,42 @@
+    }
+  });
+
+  // pairs up on the alive players
+  while (aliveUids.length > 1) {
+    // get random pair
+    const one = aliveUids[Math.floor(Math.random() * aliveUids.length)];
+    let two: string;
+    // generate a random player if the current pairs are equal
+    do {
+      two = aliveUids[Math.floor(Math.random() * aliveUids.length)];
+    } while (one === two);
+    // organize the pairs
+    pairs.push({ one, two });
+    // delete them from array
+    aliveUids.splice(aliveUids.indexOf(one), 1);
+    aliveUids.splice(aliveUids.indexOf(two), 1);
+  }
```
I think it would be better to do this something like this:
```ts
// pairs up on the alive players
while (aliveUids.length > 1) {
  // get random pair
  const [one] = aliveUids.splice(Math.floor(Math.random() * aliveUids.length), 1);
  const [two] = aliveUids.splice(Math.floor(Math.random() * aliveUids.length), 1);
  // organize the pairs
  pairs.push({ one, two });
}
```

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,160 @@
+    // Query for their chatroom
+    chatRoomInfo = await findChatRoom(lobbyData.id, user.uid);
+    // subscribe the chat messages
+    onSnapshot(
```
I think we need an unsubscribe for this

nstringham commented on client/src/components/ChatRoom.svelte:
```diff
@@ -0,0 +1,160 @@
+  }
+  // Checks if the sender is the current user
+  function isUser(uid: string) {
+    return (user as User).uid === uid;
```
looks like you don't need the `as` here

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,47 @@
+
+export async function deleteChatRooms(lobbyDoc: DocumentReference<Lobby>) {
+  const lobbyDocSnapshot = await lobbyDoc.get();
+  return db.runTransaction(async () => {
```
you are not actually using the transaction for example instead of doing
```ts
const messages = await getChatRoomMessagesCollection(room).get();
```
you should do
```ts
const messages = await transaction.get(getChatRoomMessagesCollection(room));
```
you can look at `startPrompt` or `collectPromptAnswers` in `lobby.ts` for an example of how to do it correctly or you can read the documentation here: https://firebase.google.com/docs/firestore/manage-data/transactions

gchamb committed:
refactor(client): ♻️ unsubscribe from messages

gchamb committed:
refactor(functions): ♻️ Use transaction properly and await promise all

gchamb committed:
refactor(functions): ♻️ more efficent way of getting pairs

gchamb committed:
Merge branch 'development' into giantte-muaad/chat

gchamb committed:
refactor(functions): ♻️ remove async

nstringham commented on functions/src/lobby.ts:
```diff
@@ -180,8 +188,64 @@ export const collectPromptAnswers = functions.firestore
+    // TODO: potential if checks for other states that require a timer
+    // if the state is chat then delete chatrooms
+    if (lobby.state === "CHAT") {
+      deleteChatRooms(lobbyDocRef);
```
I feel like the fact that this makes a seprate transaction might be an issue. also the duplicate lobbydoc read isn't supper ideal what if we passed `lobby` and `transaction` into this function?

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,48 @@
+        const messages = await transaction.get(getChatRoomMessagesCollection(room));
+        // get the prompt answers from the messages
+        const answers = new Map<string, string>();
+        messages.forEach((messageDoc) => {
+          // get message info
+          const message = messageDoc.data();
+          // if the message is the prompt answer then add to map
+          if (message.isPromptAnswer) {
+            answers.set(message.sender, message.text);
+          }
+          // delete all messages
+          messageDoc.ref.delete();
+        });
```
I like this

gchamb committed:
refactor(functions): ♻️ pass existing data in as a parameter

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,44 @@
+          answers.set(message.sender, message.text);
+        }
+        // delete all messages
+        messageDoc.ref.delete();
```
should this be a part of the tramsaction?

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,44 @@
+        const player = players[playerIndex];
+        player.promptAnswer = value;
+        // add back to the original array
+        players[playerIndex] = player;
```
this line doesn't do anything because when you do `player = players[playerIndex]` on line 31 it is not making a deep copy so when you do `player.promptAnswer = value` on line 32 it is modifying the players array directly

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,44 @@
+      transaction.update(lobbyDoc, { players });
+
+      // then delete room
+      room.delete();
```
this should probably be part of the transaction

nstringham commented on functions/src/lobby.ts:
```diff
@@ -180,8 +188,64 @@ export const collectPromptAnswers = functions.firestore
+    // if the state is chat then delete chatrooms
+    if (lobby.state === "CHAT") {
+      await deleteChatRooms(lobby, lobbyDocRef, transaction);
+      return transaction.update(lobbyDocRef, { state: "VOTE" });
```
I kind of don't like how this is modified here and inside of `deleteChatRooms` but I guess it's ok

gchamb committed:
refactor(functions): ♻️ delete with transaction

- delete with transaction
- move to next state within delete chatrooms
- manulating player changes

nstringham commented on functions/src/chat.ts:
```diff
@@ -0,0 +1,43 @@
+      }
+
+      // updating the players with the new players array with their answers
+      transaction.update(lobbyDoc, { players });
```
This should probably be outside the loop. Maybe merge with line 42?

nstringham replied:
For performance reasons

gchamb committed:
refactor(functions): ♻️ Update and players in same statement
